content/docs/rfcs/13/README.md

```diff
@@ -59,139 +59,260 @@ The following are not considered as part of the adversarial model:
+the client MAY make use of pagination
+in a chain of store request and response transactions
+to retrieve the full response in smaller batches termed _"pages"_.
+Pagination can be performed either in [a _forward_ or _backward_ direction](#waku-message-sorting).
+
+A store client MAY indicate the maximum number of matching entries it wants in the `StoreResponse`,
+by setting the page size limit in the `pagination_limit` field.
+Note that a store service node MAY enforce its own limit
+if the `pagination_limit` is unset
+or larger than the service node's internal page size limit.
+
+A `StoreResponse` with a populated `pagination_cursor` indicates that more stored entries match the query than included in the response.
+
+A `StoreResponse` without a populated `pagination_cursor` indicates that
+there are no more matching entries in the store.
+
+The client MAY request the next page of entries from the store service node
+by populating a subsequent `StoreRequest` with the `pagination_cursor` received in the `StoreResponse`. All other fields and query criteria MUST be the same as in the preceding `StoreRequest`.
+
+A `StoreRequest` without a populated `pagination_cursor` indicates that
+the client wants to retrieve the "first page" of the stored entries matching the query.
+
+## Store Request
+
+A client node MUST send all historical message queries within a `StoreRequest` message.
+This request MUST contain a `request_id`.
+The `request_id` MUST be a uniquely generated string.
+
+If the store client requires the store service node to include Waku message values in the query response,
+it MUST set `return_values` to `true`.
+If the store client requires the store service node to return only message hash keys in the query response,
+it SHOULD set `return_values` to `false`.
+By default, therefore, the store service node assumes `return_values` to be `false`.
+
+A store client MAY include query filter criteria in the `StoreRequest`.
+There are two types of filter use cases:
+1. Content filtered queries and
+2. Message hash lookup queries
+
+### Content filtered queries
+
+A store client MAY request the store service node to filter historical entries by a content filter.
+Such a client MAY create a filter on content topic, on time range or on both.
+
+To filter on content topic, the client MUST populate _both_ the `pubsub_topic` _and_ `content_topics` field.
+The client MUST NOT populate either `pubsub_topic` or `content_topics` and leave the other unset.
+Both fields MUST either be set or unset.
+A mixed content topic filter with just one of either `pubsub_topic` or `content_topics` set, SHOULD be regarded as an invalid request.
+
+To filter on time range, the client MUST set `time_start`, `time_end` or both.
+An unset `time_start` SHOULD be interpreted as "from the oldest stored entry".
+An unset `time_end` SHOULD be interpreted as "up to the youngest stored entry".
```

Comment:
I don't think nanosecond timestamps are possible in all languages, i remember that probably in js it is not possible. Maybe milliseconds then. But i agree we should mention the granularity.

Reply:
> Since message will have a unique hash, does it make sense to use `second` for timestamp in every places?

This is for the filter criteria for time-based queries, which doesn't rely on hashes. Maybe second granularity itself might be ok.

Reply:
> It's probably worth mentioning that time_* fields should contain unix timestamp in nanoseconds

Good point!

> I don't think nanosecond timestamps are possible in all languages

But presumably the epoch time can be converted to nanosecond timestamp with a multiplier? I think it makes sense for these fields to be the same resolution as the `timestamp` field in the Waku Message itself (which is a nanosecond Unix epoch timestamp). None of the clients can or should generate timestamp with nanosecond specificity (in fact, I would timestamp Waku messages on second or half-second boundaries) - nanoseconds is just the unit in which our timestamp fields are populated.